HTTP/2 Request Cancellation

rmkerr opened the pull request:
This PR only covers cancellation up until the point where the request body finishes sending. It covers the following steps of an HTTP/2 request:

  (1) Waiting on credit to open a stream (if MAX_CONCURRENT_STREAMS set).
  (2) Waiting on the connection write lock.
  (3) Sending headers.
  (4) Waiting on request body credit (on both the stream and the connection).
  (5) Waiting on the connection write lock again.
  (6) Sending request body.

Steps (3) and (6) will only trigger a cancellation in between frames, as cancelling the send in the middle of a frame would leave the connection in a bad state.

The change adds two additional tests for cancellation at steps (1) and (4). It's possible to test those directly since I can force the client to block and wait for a stream or a WINDOW_UPDATE, and then see what happens when I cancel the request. I've tried to cover the others situations in manual testing just by repeatedly cancelling requests after a variable amount of time, but that's more of a stress test and so I cut it out of the PR.

Fixes #31310

rmkerr commented:
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop NETFX x86 Debug Build
@dotnet-bot test Outerloop UWP CoreCLR x64 Debug Build
@dotnet-bot test Outerloop OSX x64 Debug Build

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -830,6 +830,11 @@ private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request
 
```
What about cancellation here?

geoffkizer replied:
Similarly on the FlushOutgoingBytesAsync call below.

rmkerr replied:
Specifically on FlushOutgoingBytesAsync? I decided against allowing cancellation mid-frame, as it would force us to tear the whole connection down. If you meant cancellation between serializing the headers and writing them to the wire though, I can probably plumb in cancellation there.

geoffkizer replied:
Yeah, that's a good point. We don't want to actually cancel the write to the stream.

That said, it seems that we can wait on FlushOutgoingBytesAsync indefinitely and cancellation will never kick in. That seems bad.

(It would be nice to add a test for this -- write enough data in the request body to fill the socket send queue and cause it to pend. Loopback server needs to send a large window update to allow the client to proceed, but also not process the incoming frames, so the send queue backs up. Once the send actually pends, cancel the token and verify that the operation is actually cancelled instead of hanging waiting for a send completion.)

I'm not quite sure what to do about this.

I think we need FlushOutgoingBytesAsync to be cancellable, but cancelling it will *not* cancel the stream write. Instead it will just return a cancelled task as soon as cancellation happens, but leave the connection write pending. We could do this by registering for the cancellation token and returning a cancelled Task; this gets a little messy but it seems doable.

However, I think this creates another issue. If FlushOutgoingBytesAsync is cancelled, we will end up releasing the connection writer lock, allowing someone else to take it. But at this point, the pending write may still not be complete. I think that means we need to track whether we have a pending send or not, and if so, await it when we try to take the writer lock. 

Thoughts?

geoffkizer replied:
Just to be clear, I think this corresponds to steps (3) and (6) in your list above, correct?

rmkerr replied:
Yeah, this does correspond to steps (3) and (6).

Just to make sure we're on the same page: you're suggesting we minimize the time between when a client cancels the CancellationToken and when we complete the TaskCompletionSource with an exception.

We currently only handle FlushOutgoingBytesAsync as if it were atomic -- we'll cancel before or after it completes, but not during it. If we implement what you suggested, we'll be able to report a cancellation to whoever initiated the request during FlushOutGoingBytesAsync. We won't actually cancel the write though, and will still block the connection along with any other pending requests.

I think the implementation you suggested seems doable, if the benefits here outweigh the additional complexity.

geoffkizer replied:
I wouldn't use the term "minimize". In general it's fine for cancellation to take some time to propagate. 

What's not okay, it seems to me, is for cancellation to be unable to happen because of external factors we have no control over. That seems to be the case here. If the server doesn't process even unblock our socket send queue, cancellation will never happen.

rmkerr replied:
Okay, I think that's reasonable. I'll try to get this working in the next iteration.

rmkerr replied:
This should be working in the latest commit, along with a new test that checks this behavior.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1150,6 +1167,15 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
+                else if (e is OperationCanceledException)
+                {
+                    // If the operation has been cancelled after the stream was allocated an ID, send a RST_STREAM.
+                    if ( http2Stream != null && http2Stream.StreamId != 0 )
```
remove spaces here

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -74,12 +76,22 @@ public async Task SendRequestBodyAsync()
+                        }
+                        catch (OperationCanceledException)
+                        {
+                            // We must handle cancellation before disposing the stream, so that we can make a decision about
```
This seems redundant with the catch in the caller

rmkerr replied:
The issue here is the `using` statement. Http2WriteStream.Dispose currently sends an empty data frame with the END_STREAM header set, unless the request was cancelled. So if a cancellation occurs during CopyToAsync we'll send the END_STREAM header and then send a RST_STREAM. This code makes sure we cancel first, and don't send the extra frame.

geoffkizer replied:
Ah, I understand. Hmm, let me think about this.

rmkerr replied:
That sounds reasonable to me. I'll give it a try, and see if things end up cleaner.

rmkerr replied:
This is done in the latest commit.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -331,6 +343,18 @@ public void Dispose()
+                    if(!_disposed)
+                    {
+                        _cancelled = true;
+                        ValueTask ignored = _connection.SendRstStreamAsync(_streamId, Http2ProtocolErrorCode.StreamClosed);
```
I don't think StreamClosed is the correct error code to send. InternalError might be best; this indicates that the cause of the RST_STREAM is internal to the client.

rmkerr replied:
Good catch. Looking at the spec, there's a dedicated error code for this called CANCEL that I think is appropriate:

>If the client determines, for any reason, that it does not wish to receive the pushed response from the server or if the server takes too long to begin sending the promised response, the client can send a RST_STREAM frame, using either the CANCEL or REFUSED_STREAM code and referencing the pushed stream's identifier.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -830,6 +830,11 @@ private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request
                 {
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        throw new OperationCanceledException(cancellationToken);
```
I wonder if it's actually legal to send RST_STREAM in the middle of headers (i.e. after one or more HEADERS/CONTINUATION frames but before the EndHeaders flag is sent). What does the spec say here?

rmkerr replied:
Ah, that's a reasonable behavior. I'll remove this.

geoffkizer replied:
This does complicate the question of cancellation and how it interacts with FlushOutgoingBytesAsync, because it means we either need to cancel before we send HEADERS at all, or we need to send the full HEADERS and CONTINUATION blocks, while still allowing the request to be cancelled while the send of those is in progress...

geoffkizer commented on src/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.Http2.cs:
```diff
@@ -806,5 +807,105 @@ public async Task Http2_MaxConcurrentStreams_LimitEnforced()
+
+                cts.Cancel();
+
+                await Assert.ThrowsAsync<OperationCanceledException>(async () => await sendTask);
```
Do we need to ensure that the Task is cancelled promptly here? I suppose that the only two possible outcomes are either cancelled promptly or not cancelled at all, is that right? Even so, it might be a good idea to ensure that it's actually cancelled promptly.

Similarly for the test below.

rmkerr replied:
We could, but I'm hesitate to add additional time sensitive tests that are likely to be flaky on ARM.

rmkerr replied:
Actually, I'm going to revisit this. Given that the HTTP/2 loopback will already time out if frames arrive extremely slowly, I think this is okay. I'll update the tests.

geoffkizer commented:
Have you looked at the existing HTTP/1.1 cancellation tests?

Ideally we would adapt those to use the generic loopback server and run them for HTTP2 as well. But since that's not available yet, we can't do that right now. And I'm not sure how amenable those tests actually are to being made version-agnostic.

That said, seems like we want to get to a place where we have:
(1) Some version-agnostic tests, where possible
(2) Some HTTP/1.1 specific tests, if necessary and appropriate
(3) Some HTTP2 specifc tests that test for cancellation at specific given points in the HTTP2 request handling (i.e. the tests you added here)

Overall I just want to make sure that we have sufficient test coverage here.

geoffkizer commented:
Also re stress testing: Do we have any existing stress tests for cancellation? I vaguely remember @stephentoub adding something like this. It would probably be good to add similar HTTP2 tests.

rmkerr commented:
I've just pushed up a PR addressing some of these comments, just to get the tests running. Feel free to hold off on reviewing until I address all of the comments.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/CreditManager.cs:
```diff
@@ -47,14 +49,18 @@ public ValueTask<int> RequestCreditAsync(int amount)
                 }
 
-                var tcs = new TaskCompletionSource<int>(TaskContinuationOptions.RunContinuationsAsynchronously);
+                var tcs = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
```
Good catch.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/CreditManager.cs:
```diff
@@ -47,14 +49,18 @@ public ValueTask<int> RequestCreditAsync(int amount)
-                _waiters.Enqueue(new Waiter { Amount = amount, TaskCompletionSource = tcs });
+                Waiter waiter = new Waiter { Amount = amount,
+                                             TaskCompletionSource = tcs,
+                                             TokenRegistration = cancellationToken.Register(() => tcs.TrySetCanceled(cancellationToken)) };
```
The allocation here is slightly tricky to avoid in this case, because you need to pass in both the tcs and the cancellation token.  But that issue was already encountered and solved, via https://github.com/dotnet/corefx/blob/f7539b726c4bc2385b7f49e5751c1cff2f2c7368/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L1041.  So you'd want to make your TaskCompletionSource property actually be of that type, and then here have the registration be:
```C#
cancellationToken.Register(s => ((TaskCompletionSourceWithCancellation<int>)s).OnCanceled());
```

rmkerr replied:
Thanks. I was struggling to figure out how to do this.

geoffkizer replied:
If I remember correctly, TaskCompletionSourceWithCancellation actually handles the token Register and Dispose for you. So you can just replace the line below with:

return new ValueTask<int>(tcs.WaitWithCancellationAsync(cancellationToken));

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/CreditManager.cs:
```diff
@@ -106,7 +118,8 @@ public void Dispose()
                     {
-                        waiter.TaskCompletionSource.SetException(new ObjectDisposedException(nameof(CreditManager)));
+                        waiter.TaskCompletionSource.TrySetException(new ObjectDisposedException(nameof(CreditManager)));
+                        waiter.TokenRegistration.Dispose();
```
Swap these lines to dispose of the registration first?

rmkerr replied:
With the change to `TaskCompletionSourceWithCancellation`, I've been able to remove this line altogether. So this should be resolved in the next commit.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -627,7 +627,7 @@ private void ReleaseWriteLock()
         private async ValueTask SendSettingsAckAsync()
         {
-            await AcquireWriteLockAsync().ConfigureAwait(false);
+            await AcquireWriteLockAsync(CancellationToken.None).ConfigureAwait(false);
```
FWIW, you can also do `default`.  They're the same, since None will be inlined and return default, so whichever you prefer stylistically.  We should just be consistent throughout.

stephentoub replied:
AcquireWriteLockAsync could also be defined to have its `CancellationToken cancellationToken = default`, i.e. an optional argument, in which case the call site wouldn't need to be modified in this case.

rmkerr replied:
I've updated the change to use `default`, so this should be fixed in the next commit.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -851,20 +851,32 @@ private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request
+                {
+                    await AcquireWriteLockAsync(cancellationToken).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
```
Is this specific to cancellation?  Or would we do that for any exception?

geoffkizer replied:
Any exception here other than cancellation should cause the connection to be killed, so I don't think it matters.

That said, I think this is probably related to the question of how we handle cancellation in FlushOutgoingBytesAsync....

geoffkizer replied:
I should add, it's a little odd that we have the try/finally below where we call ReleaseWriterLock in the finally. If we throw in that block, the connection is going to be killed, as mentioned above. But we still need to ensure writers are unblocked, even though they are going to fail their writes.

An alternative here would be to just dispose the writer lock in FlushOutgoingBytesAsync if a write exception occurs. That would cause any waiting writers to throw, which is probably slightly better than having them unblock, try to write, and then fail. And we wouldn't need the try/finally here.

rmkerr replied:
I think that the specific try-catch referenced here is necessary. Cancellation at this point is not necessarily going to kill the connection, so we need to release the credit acquired by this request. Otherwise we might end up hanging if our credit accounting differs from the servers.

I don't have an issue removing the try-finally block below. It does mean that any exceptions during request serialization will be fatal to the connection, but that seems like a reasonable behavior.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -34,6 +34,7 @@ sealed class Http2Stream : IDisposable
             private readonly HttpResponseMessage _response;
             private readonly HttpConnectionResponseContent _responseContent;
             private bool _disposed;
+            private bool _cancelled;
```
Nit: cancel, canceling, canceler, canceled, cancellation... i.e. two Ls in cancellation, one L in everything else.  We're not always consistent on those spellings, but we try :)

rmkerr replied:
Even though I know this, I always instinctively add two 'l's. I'll go over this before the next commit :)

geoffkizer commented:
Re canceling FlushOutgoingBytesAsync and related issues from above:

A related issue here is how we manage flushing of the write buffer. Currently we force it to be flushed on every frame. This is unnecessarily aggressive. 

What we should do instead, I think, is write as many frames into the buffer as we can, until we run out of buffer space, then flush the buffer. And further, even when there is not enough room for the whole frame, we could choose to write part of a frame into the buffer, flush, and then write the rest of the frame.

We do need to aggressively send frames sometimes, to ensure the server receives them promptly. However, this is never true for headers or request body, since there will always be more frames to follow (or eventually, a DATA frame with EndStream or a RST_STREAM frame). We do want to aggressively flush EndStream or RST_STREAM frames, as well as protocol-level frames like WINDOW_UPDATE or SETTINGS ack etc. But even here, we only need to do this if there are no other writers pending, because if there are, we may as well consolidate their writes into the same buffer too. (Assuming they aren't canceled, of course...)

I'm not quite sure how this interacts with cancellation specifically. I guess what I'm saying is, we may want to look at how to rework AcquireWriteLockAsync/FlushOutgoingBytesAsync in general, and if we do, we can thread cancellation through this properly.

geoffkizer commented:
> Currently we force it to be flushed on every frame. This is unnecessarily aggressive.

Just to illustrate why this matters:

Today, if you do a POST, we will end up doing at least three writes: One for the HEADERS frame, one for the DATA frame, and one for the final empty DATA frame with EndStream set. In contrast, for HTTP/1.1 we can do this all in one write if it fits in the connection write buffer. I believe this will have a noticeable perf impact.

geoffkizer commented:
> we may want to look at how to rework AcquireWriteLockAsync/FlushOutgoingBytesAsync in general

Here's a stab at how this would work:

(1) Replace AcquireWriterLockAsync with StartWriteAsync(int spaceNeeded), which also replaces EnsureSpaceAvailable.
(2) Replace ReleaseWriterLock with FinishWrite(bool flushNeeded), which also removes the need for FlushOutgoingBytesAsync().

FinishWrite will decide whether to flush or not. If flushNeeded is true (or a previous FinishWrite set it to true), and there are no pending writers, then it will perform the write asynchonously and store the resulting Task on the connection. At this point, we don't care if it pends. It will then release the writer lock.

StartWriteAsync will acquire the writer lock, and first check if there is an outstanding write Task. If so, await it and reset the buffer (and possibly grow it if needed). If not, see if there is sufficient buffer space; if not, flush and await and/or grow the buffer as appropriate. At the same time, use TaskCompletionSourceWithCancellation to allow the caller to be canceled even while we are awaiting the write. The write will be completed eventually even if the caller cancels.

This ensures that the only async operation here is StartWriteAsync, and it is fully cancellable. 

This approach doesn't allow for splitting frames (or HEADERS/CONTINUATION since they must be contiguous) across writes. But it's pretty ugly to support that in a cancellable way, because once we've performed a write of a partial frame, we need to complete it somehow to keep the connection valid, and so that requires us to not cancel. As such, this seems like a reasonable tradeoff to me.

rmkerr commented:
> (1) Replace AcquireWriterLockAsync with StartWriteAsync(int spaceNeeded), which also replaces EnsureSpaceAvailable.
(2) Replace ReleaseWriterLock with FinishWrite(bool flushNeeded), which also removes the need for FlushOutgoingBytesAsync().

Everywhere we use AcquireWriteLockAsync and ReleaseWriterLock follows almost the exact same pattern. For example:
https://github.com/dotnet/corefx/blob/47f6404357832466585f06d5c9ee07ed2f7a27a4/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L628-L642

The suggested changes simplify that quite nicely, while also allowing us to hide the new `FlushOutgoingBytesAsync` cancellation logic in `StartWriteAsync`:
```csharp
private async ValueTask SendSettingsAckAsync()
{
    await StartWriteAsync(FrameHeader.Size); // Note [1]
    try
    {
        WriteFrameHeader(new FrameHeader(0, FrameType.Settings, FrameFlags.Ack, 0));
    }
    finally
    {
        await FinishWrite(false);
    }
}
```

[1] In the interests of keeping error handling simple, StartWriteAsync will need to release `_writerLock` if `EnsureSpaceAvailable` or the potential outstanding write throws.

This is the simplest example, but updating `SendHeadersAsync` should follow roughly the same pattern.

Since this simplifies the other work around cancelling `FlushOutgoingBytesAsync` and doesn't add a lot of complexity to the rest of the code, I think it's worth making that change in this PR.

rmkerr commented:
> If so, await it and reset the buffer (and possibly grow it if needed). If not, see if there is sufficient buffer space; if not, flush and await and/or grow the buffer as appropriate.

Since `EnsureAvailableSpace` just doubles the size of the buffer if we don't have enough space (and so never really "fills up"), we'll need to make a decision about when to expand the buffer vs flushing it. Do you have a heuristic in mind for how we should decide that?

geoffkizer commented:
Off the top of my head: Grow up to 32K, then flush instead of growing.

rmkerr commented:
There's still some WIP stuff in here, so I'm going to re-add that tag. Otherwise, the latest commits should include most of the fixes and improvements discussed in the last iteration.

rmkerr commented:
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop NETFX x86 Debug Build
@dotnet-bot test Outerloop UWP CoreCLR x64 Debug Build
@dotnet-bot test Outerloop OSX x64 Debug Build

rmkerr commented:
/azp run corefx-outerloop-windows

rmkerr commented:
/azp run corefx-outerloop-linux

rmkerr commented:
/azp run corefx-outerloop-osx

rmkerr commented:
/azp run corefx-outerloop-windows

rmkerr commented:
/azp run corefx-outerloop-linux

rmkerr commented:
/azp run corefx-outerloop-osx

rmkerr commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -622,80 +629,128 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
+                    
+                    var tcs = new TaskCompletionSourceWithCancellation<bool>();
+                    
+                    if (_inProgressWrite != await Task.WhenAny(_inProgressWrite, tcs.WaitWithCancellationAsync(cancellationToken)).ConfigureAwait(false))
```
I would appreciate some additional scrutiny of this code, since it's the first time I've tried to implement something like this. I based this off of the information [here](https://blogs.msdn.microsoft.com/pfxteam/2012/10/05/how-do-i-cancel-non-cancelable-async-operations/).

geoffkizer replied:
I don't think we need to wait here at all. We've written the buffer to the kernel's send buffer. We don't care at this point if this completes. We just need to wait for it in StartWriteAsync if it hasn't completed.

Combined with my comment above, I think this means you can (a) remove the cancellation token here and (b) make this function synchronous.

geoffkizer replied:
To clarify: we haven't necessarily written it to the kernel buffer yet, but that doesn't matter, because we've called WriteAsync on the underlying stream so it will be written at some point. We don't care when the actual write happens. We just need to wait on it in StartWriteAsync, and deal with failures there.

rmkerr replied:
> We've written the buffer to the kernel's send buffer.

Is that always going to be true? I know that we have called `SslStream.WriteAsync`, but we haven't awaited it. Will it always copy the data over into the kernel send buffer before returning?

rmkerr replied:
Actually, let me revise that. I see what you mean here, and will go ahead and update this section of code.

rmkerr commented:
/azp run corefx-outerloop-windows

rmkerr commented:
/azp run corefx-outerloop-linux

rmkerr commented:
/azp run corefx-outerloop-osx

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -622,80 +629,128 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
         }
 
-        private async ValueTask AcquireWriteLockAsync()
+        private async ValueTask StartWriteAsync(int credit, CancellationToken cancellationToken = default, bool acquireWriteLock = true)
```
It's confusing to call this parameter "credit" -- I originally assumed this had to do with window size management. Maybe call it "writeBytes" or something like that?

rmkerr replied:
Thanks. I've fixed this and plan to include the change in the next iteration.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -622,80 +629,128 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
+                int activeBufferLength = _outgoingBuffer.ActiveSpan.Length;
 
-            _writerLock.Release();
+                if (totalBufferLength >= 32768 &&
```
We should never have "dead" space at the beginning of the buffer here. The pattern should always be: Fill the buffer until we are ready to flush it, then flush the entire buffer (up to the end of the active bytes), then discard all active bytes, meaning the buffer is now empty and we start writing at the beginning again.

This is somewhat obscured by the fact that we use a single structure (ArrayBuffer) for both read buffer and write buffer. In retrospect I'm not sure this is a great idea. 

I don't think this actually changes the code here, just FYI.

rmkerr replied:
I can add a comment explaining this in the code, if you think this needs some additional clarity. Otherwise I'll leave it as-is.

geoffkizer replied:
It's fine. More just food for thought, in case we decide to simplify the buffer management here.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -897,25 +953,27 @@ private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request
                     flags = (remaining.Length == 0 ? FrameFlags.EndHeaders : FrameFlags.None);
 
-                    _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size + current.Length);
+                    await StartWriteAsync(FrameHeader.Size + current.Length, default, false).ConfigureAwait(false);
```
I think we are still not handling write atomicity properly here. We could send a HEADERS block, then cancel before sending the CONTINUATION.

My suggestion here is to just write this all out in a single StartWriteAsync pass. I think that will also remove the need for the acquireWriteLock flag on StartWriteAsync, which will simplify things a bit.

This means we need a separate semaphore to serialize stream creation, but that seems fine. There's already a comment above saying we may want to do this, and it will probably reduce write lock contention in some scenarios.

rmkerr replied:
> My suggestion here is to just write this all out in a single StartWriteAsync pass. I think that will also remove the need for the acquireWriteLock flag on StartWriteAsync, which will simplify things a bit.

The code change here is simple enough, but it does mean that the write buffer can grow larger than the maximum header frame size in the case of super large headers. That's a very unusual case though, so I'm not sure we need to care about it.

geoffkizer replied:
Yeah, I don't think we need to worry about that.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -897,25 +953,27 @@ private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request
                     _outgoingBuffer.Commit(current.Length);
 
-                    await FlushOutgoingBytesAsync().ConfigureAwait(false);
+                    await FinishWriteAsync(remaining.Length == 0, cancellationToken, false).ConfigureAwait(false);
```
I think you only need to force the flush in the EndStream case. If it's not EndStream then there will be body frames to follow.

rmkerr replied:
This will be fixed in the next iteration. Thanks.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -622,80 +629,128 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
+                // We must flush if the caller requires it, or if there are no other pending writes.
+                if (mustFlush || (releaseWriteLock && _pendingWriters == 0))
+                {
+                    if (_inProgressWrite != null)
```
How can this be non-null here? It seems like we should have waited for it in StartWriteAsync, no?

rmkerr replied:
You're right. This is left over from my first pass at implementing these, and it no longer applies. I've replaced it with a debug assert in the next commit:
```csharp
Debug.Assert(_inProgressWrite == null);
```

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -360,6 +363,14 @@ public void Dispose()
                 }
             }
 
+            public void Cancel()
```
How does the Http2Stream get removed from the dictionary in this case?

rmkerr replied:
It currently isn't removed here, but it should be. I'll make that change.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -360,6 +363,14 @@ public void Dispose()
+                lock (SyncObject)
+                {
+                    Task ignored = _connection.SendRstStreamAsync(_streamId, Http2ProtocolErrorCode.Cancel);
+                }
```
What about _streamState here?

rmkerr replied:
I've gone ahead and updated this, but I'm still trying to wrap my head around the other changes that came in with the same PR. Will I need to update this to follow the same pattern as OnResponseAbort where I need to signal the next waiter?

geoffkizer replied:
I put some thoughts on this and related issues in #35652. Thoughts on that?

rmkerr replied:
I've added some thoughts on the linked issue. I think we should probably add some additional values to StreamState (i.e, Canceled) when we make the changes described there but for now I've set it as Aborted.

geoffkizer commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs:
```diff
@@ -343,6 +343,36 @@ public async Task MaxConnectionsPerServer_WaitingConnectionsAreCancelable()
         }
 
+        [Fact]
+        public async Task SendAsync_Cancel_CancellationTokenPropagates()
```
Is the intent to run this test for HTTP2? I don't see where these tests are running for HTTP2 currently.

Also, what about the other cancellation tests here?

rmkerr replied:
That was my intent. Let me try to update that.

The other tests here don't abstract from the 1.1 loopback server to the 2.0 loopback server, at least with the current API.

rmkerr replied:
This is fixed in the latest iteration.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1278,6 +1326,15 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
+                    // If the operation has been canceled after the stream was allocated an ID, send a RST_STREAM.
+                    if (http2Stream != null && http2Stream.StreamId != 0)
+                    {
+                        http2Stream.Cancel();
```
There's actually a TODO in Http2Stream.Dispose, to add something like this. See: https://github.com/dotnet/corefx/blob/694eeba312046e2097c731d40f32c1cac675aa03/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L358

Note we call Dispose just above in this code. I think Dispose is probably the right place to do this, because we will need to handle this in other exception cases too, e.g. receiving an invalid HEADERS block or something like that.

geoffkizer replied:
BTW, I was looking at adding some tests for receiving invalid HEADERS blocks. We are definitely not handling this correctly today, and need to do something smarter. Sending the RST_STREAM is somewhat independent of this PR, since we need a way to handle a variety of exceptional cases that will cause RST_STREAM; cancellation is just one of these.

geoffkizer replied:
I just added #35652 with a bunch of thoughts about this.

rmkerr replied:
I'll leave this for a separate PR, if that's okay with you.

geoffkizer replied:
That's fine.

rmkerr commented:
/azp run

rmkerr commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -867,56 +921,78 @@ private void WriteHeaders(HttpRequestMessage request)
+            ReadOnlyMemory<byte> current = null;
+            FrameFlags flags;
 
             try
```
I need to handle cancellation differently when we're serializing the headers vs when we're actually writing to the wire. Splitting this up into two separate try-catch blocks feels a little bit awkward, but I wasn't sure how else to handle it. Thoughts?

geoffkizer replied:
I don't think we can release the header lock before we do the write. Part of what it is protecting is that the stream IDs in headers blocks must always increase. If we release the header lock before the writer lock, we could do the writes out of order.

rmkerr replied:
Thanks. I'll fix that. I think it will actually simplify things a bit here.

rmkerr commented:
/azp run

rmkerr commented:
/azp run

faskowbn commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/TaskCompletionSourceWithCancellation.cs:
```diff
@@ -0,0 +1,32 @@
+            }
+        }
+    }
+}
```
Have you considered adding a newline at EOF?

rmkerr replied:
Added, thanks.

faskowbn commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -867,28 +921,33 @@ private void WriteHeaders(HttpRequestMessage request)
                 Debug.Assert(remaining.Length > 0);
 
+                // Calculate the total number of bytes we're going to use (content + headers).
+                int totalSize = remaining.Length + (remaining.Length / FrameHeader.MaxLength) * FrameHeader.Size +
```
If you are going to calculate it this way, I'd recommend adding parens to be explicit about order of mathematical ops

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/CreditManager.cs:
```diff
@@ -55,16 +56,19 @@ public ValueTask<int> RequestCreditAsync(int amount)
-                return new ValueTask<int>(tcs.Task);
+                _waiters.Enqueue(waiter);
+
+                return new ValueTask<int>(tcs.WaitWithCancellationAsync(cancellationToken));
```
I'd suggest making this:
```C#
return new ValueTask<int>(cancellationToken.CanBeCanceled ?
    tcs.WaitWithCancellationAsync(cancellationToken) :
    tcs.Task;
```
Otherwise, if the cancellation token might have been `CancellationToken.None`/`default(CancellationToken)`, we'll end up paying additional unnecessary costs (e.g. an extra async method state machine).

rmkerr replied:
I'll fix this for the next commit. Thanks.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -622,80 +631,125 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
+                // If there is a pending write that was canceled while in progress, wait for it to complete.
+                if (_inProgressWrite != null)
+                {
+                    await _inProgressWrite;
```
.ConfigureAwait(false)

rmkerr replied:
Argh. Fixed.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -622,80 +631,125 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
+                    // and is non-empty, flush the current contents to the wire.
+                    await FlushOutgoingBytesAsync().ConfigureAwait(false);
+                }
+                _outgoingBuffer.EnsureAvailableSpace(writeBytes);
```
Nit: add a blank line above this one

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -622,80 +631,125 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
+                int activeBufferLength = _outgoingBuffer.ActiveSpan.Length;
 
-            _writerLock.Release();
+                if (totalBufferLength >= 32768 &&
```
Where did this 32768 number come from?  Is that tied to something else?  Configurable?  Assuming it's something we can dial and/or that it's tied to something else in the code, we should make it a const at the top of the class.  You might also consider defining it as `32 * 1024` to make it clear what the number is (though most folks reading the code will probably recognize it).

rmkerr replied:
It's pretty much just an arbitrary value. I'll move it to the top of the class, and add a comment describing what it means.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -622,80 +631,125 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
 
-        private async ValueTask SendSettingsAckAsync()
+        // This method handles flushing bytes to the wire. Writes here need to be atomic, so as to avoid
+        // killing the whole connection. Callers must hold the write lock, but can specify whether or not
```
This won't validate that the actual caller holds the write lock, but you could at least do `Debug.Assert(_writerLock.CurrentCount == 0)` to check that _someone_ holds the write lock.

rmkerr replied:
Sounds reasonable. I'll add this.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -622,80 +631,125 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
+                    
+                    var tcs = new TaskCompletionSourceWithCancellation<bool>();
+                    
+                    if (_inProgressWrite != await Task.WhenAny(_inProgressWrite, tcs.WaitWithCancellationAsync(cancellationToken)).ConfigureAwait(false))
```
This is problematic in that until cancellationToken has cancellation requested (which may never happen), it's going to leak a registration from WaitWithCancellationAsync into the cancellation token.

At a minimum you should call tcs.TrySetResult(true), which will cause the WaitWithCancellationAsync method to complete and unregister from the token.

But this is also a fairly expensive call in general, especially since it's _inProgressWrite will already be completed.  The whole creation of the tcs and the if block should be surrounded by a check:
```C#
if (!_inProgressWrite.IsCompleted)
{
    ...
}
```
such that we only pay these costs if we actually have an asynchronously-completing write that hasn't yet completed.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -622,80 +631,125 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
-                _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size);
-                WriteFrameHeader(new FrameHeader(0, FrameType.Settings, FrameFlags.Ack, 0));
+                // We must flush if the caller requires it, or if there are no other pending writes.
+                if (mustFlush || (releaseWriteLock && _pendingWriters == 0))
```
What happens if there are pending writers but we read _pendingWriters as 0?  That appears to be possible given the synchronization involved; presumably if that happens it just means we flush when we wouldn't otherwise need to, and so worst case it's just a perf impact but not a functional impact, correct?

rmkerr replied:
That's correct. It's acceptable (though not ideal) to flush unnecessarily. The goal here was to err on the side of flushing, since failing to flush when there are no pending writers could cause us to never actually write the bytes to the wire.

If you see any potential ways to make this smarter about flushing though, don't hesitate to let me know.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -622,80 +631,125 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
             {
-                ReleaseWriteLock();
+                Interlocked.Decrement(ref _pendingWriters);
+            }
```
Could we write this instead as:
```C#
Task acquireLockTask = _writerLock.WaitAsync(cancellationToken);
if (!acquireLockTask.IsCompletedSuccessfully)
{
    Interlocked.Increment(ref _pendingWriters);
    try { await acquireLockTask.ConfigureAwait(false); }
    finally { Interlocked.Decrement(ref _pendingWriters);
}
```
such that we only pay the additional interlocked costs if we didn't successfully acquire the lock already?

rmkerr replied:
I've reworked the code to follow this pattern.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -622,80 +631,125 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
         }
 
-        private async Task SendRstStreamAsync(int streamId, Http2ProtocolErrorCode errorCode)
+        private async ValueTask SendSettingsAckAsync()
```
This can/should return Task instead of ValueTask; there's no benefit to ValueTask here.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -622,80 +631,125 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
+            WriteFrameHeader(new FrameHeader(0, FrameType.Settings, FrameFlags.Ack, 0));
 
-                BinaryPrimitives.WriteInt32BigEndian(_outgoingBuffer.AvailableSpan, (int)errorCode);
+            await FinishWriteAsync(true).ConfigureAwait(false);
```
Nit: can you use a named argument for the `true` here?  It's not clear at the call site otherwise what `true` means.

rmkerr replied:
Sure.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -622,80 +631,125 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
+        }
 
-                _outgoingBuffer.Commit(FrameHeader.RstStreamLength);
+        private async ValueTask SendPingAckAsync(ReadOnlyMemory<byte> pingContent)
```
Same as above; use Task instead of ValueTask here.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -622,80 +631,125 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
+            pingContent.CopyTo(_outgoingBuffer.AvailableMemory);
+            _outgoingBuffer.Commit(FrameHeader.PingLength);
+
+            await FinishWriteAsync(false).ConfigureAwait(false);
```
Same comment about using a named argument for the bool arg to FinishWriteAsync.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -867,28 +921,33 @@ private void WriteHeaders(HttpRequestMessage request)
-            await _writerLock.WaitAsync().ConfigureAwait(false);
+            // We serialize usage of the header encoder and the header buffer seperately from the
+            // write lock
+            await _headerSerializationLock.WaitAsync(cancellationToken).ConfigureAwait(false);
```
It's hard for me to tell... do we have a strict ordering as to the acquisition of _headerSerializationLock and _writerLock?  We need to make sure that if one can be held while acquiring the other that we always do so in the same order; otherwise we risk deadlock.

stephentoub replied:
Also, can you clarify why a separate lock is needed?  Why can't it just hold the single write lock across both operations?

rmkerr replied:
There is a strict ordering to the locks here. _headerSerializationLock is only needed for this function, and will always be acquired before _writerLock.

The goal of the _headerSerializationLock is to protect the header buffer, and other associated resources. We can only have a single thread accessing those resources at a time, and it is not safe to release that lock until that buffer has been flushed to the write buffer, which is why we end up acquiring both locks here.

On the other hand, it is actually safe for threads to write other (non header) data to the write buffer while we're serializing headers. As long as they don't touch the header, that's a-ok. Having a seperate lock allows us to do that.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -867,28 +921,33 @@ private void WriteHeaders(HttpRequestMessage request)
-            // (If necessary, we could have a separate semaphore just for creating and encoding header blocks,
-            // and defer taking the actual _writerLock until we're ready to do the write below.)
-            await _writerLock.WaitAsync().ConfigureAwait(false);
+            // We serialize usage of the header encoder and the header buffer seperately from the
```
Nit: seperately => separately

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -926,77 +986,67 @@ private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request
         }
 
-        private async ValueTask SendStreamDataAsync(int streamId, ReadOnlyMemory<byte> buffer)
+        private async ValueTask SendStreamDataAsync(int streamId, ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
```
ValueTask => Task

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -926,77 +986,67 @@ private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request
+            await FinishWriteAsync(true).ConfigureAwait(false);
         }
 
         private async ValueTask SendWindowUpdateAsync(int streamId, int amount)
```
ValueTask => Task

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -926,77 +986,67 @@ private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request
             }
         }
 
         private async ValueTask SendEndStreamAsync(int streamId)
```
ValueTask => Task

For things that don't affect public surface area, for methods implemented with async/await, default to using:
- `Task` if the method doesn't return a value
- `ValueTask<T>` if the method does return a value

rmkerr replied:
I've gone ahead and updated all of the instances you commented on, and a couple of other locations based on what you suggested here. Thanks.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -926,77 +986,67 @@ private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request
-            {
-                ReleaseWriteLock();
-            }
+            WriteFrameHeader(new FrameHeader(0, FrameType.Data, FrameFlags.EndStream, streamId));
```
Is there any situation in which WriteFrameHeader could throw?  If it could, it seems like we'd end up deadlocking because we'd never release the lock?

rmkerr replied:
We have some Debug.Asserts that could fail in WriteFrameHeader, but I don't think that we can throw otherwise.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1285,6 +1335,15 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
+                    {
+                        http2Stream.Cancel();
+                    }
+                    throw new OperationCanceledException(cancellationToken);
```
Should we instead just propagate the original exception?
```C#
throw;
```
Or is it possible the original exception might not be associated with this cancellation token?  In which case:
```C#
else if (e is OperationCanceledException oce)
{
    ...

    if (oce.CancellationToken == cancellationToken)
    {
        throw;
    }

    throw new OperationCanceledException(cancellationToken);
}
```
?

rmkerr replied:
I've tried to be consistent about associated the CancellationToken, but there are some places where I'm depending on other components throwing. I'll opt for the second approach, just to be safe.

rmkerr replied:
This actually surfaced a different issue: TrySetCancelled throws TaskCanceledException, but I've setting things up to throw and expect OperationCanceledException. It isn't totally clear to me which we should be throwing, as our behavior throughout SocketsHttpHandler doesn't seem to be consistent. Any thoughts on this?

stephentoub replied:
TaskCanceledException derives from OperationCanceledException... it's just an OperationCanceledException that also carries with it the Task that was canceled.  In general developers should be catching the base OCE, not the derived TCE.  However, we've tried to always throw TCE whenever it's appropriate: https://github.com/dotnet/corefx/pull/28075

rmkerr replied:
Okay. I'll update this to follow the same pattern.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -394,18 +397,18 @@ public async ValueTask<int> ReadDataAsync(Memory<byte> buffer, CancellationToken
             }
 
-            private async ValueTask SendDataAsync(ReadOnlyMemory<byte> buffer)
+            private async ValueTask SendDataAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
```
ValueTask => Task

rmkerr replied:
The only usage of this function is in the implementation of `WriteAsync`, which is defined in the base class `BaseAsyncStream`, and returns ValueTask. I can rework that code to use Task instead of ValueTask if you'd like, but that seems like it's out of the scope of this PR.

stephentoub replied:
> but that seems like it's out of the scope of this PR

Ok.  We can fix it up separately.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/TaskCompletionSourceWithCancellation.cs:
```diff
@@ -0,0 +1,32 @@
+        public async Task<T> WaitWithCancellationAsync(CancellationToken cancellationToken)
+        {
+            _cancellationToken = cancellationToken;
+            using (cancellationToken.Register(s => ((TaskCompletionSourceWithCancellation<T>)s).OnCancellation(), this))
```
Nit: Register => UnsafeRegister.  The latter doesn't flow ExecutionContext into the callback; we don't need the ExecutionContext captured/flowed here (we control all of the code in the callback and know exactly what it does), and avoiding it is a small optimization.

rmkerr replied:
Thanks. This should be fixed in the latest commit.

rmkerr commented:
/azp run

rmkerr commented:
There were some infrastructure related failures here, but otherwise the test run was okay. I'm going to re-run the tests, but I think this is ready for another round of review.

/azp run

rmkerr commented:
Bump. This is ready for additional review.

karelz commented:
@stephentoub can you please do final review round? Thanks!

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ArrayBuffer.cs:
```diff
@@ -59,6 +59,8 @@ public void Dispose()
         public Memory<byte> ActiveMemory => new Memory<byte>(_bytes, _activeStart, _availableStart - _activeStart);
         public Memory<byte> AvailableMemory => new Memory<byte>(_bytes, _availableStart, _bytes.Length - _availableStart);
 
+        public int TotalLength => _bytes.Length;
```
Nit: Often this is called Capacity

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -622,80 +636,116 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
+                {
+                    await _inProgressWrite.ConfigureAwait(false);
+                    _inProgressWrite = null;
+                }
```
What happens if an in-progress write fails.  Could that cause every subsequent write attempt to get here, see that _inProgressWrite is non-null, and then await it and throw the same exception, with no one ever getting past this point?  I'm wondering if instead this should be:
```C#
Task inProgressWrite = _inProgressWrite;
if (inProgressWrite != null)
{
    _inProgressWrite = null;
    await inProgressWrite.ConfigureAwait(false);
}
```

rmkerr replied:
The situation you described is possible, but I think it's actually the desired behavior. Once a write fails we're going to have to kill the whole connection, so no other requests should get through. The first exception thrown should eventually lead to the connection being torn down, so we won't be stuck in this state indefinitely.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -867,28 +917,33 @@ private void WriteHeaders(HttpRequestMessage request)
+            // write lock
+            await _headerSerializationLock.WaitAsync(cancellationToken).ConfigureAwait(false);
 
             Http2Stream http2Stream = AddStream(request);
```
Is it possible this could throw?  If it does, we could leave the _headerSerializationLock held.

rmkerr replied:
It is. I'll move this into the try block.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1285,6 +1331,22 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
+                        http2Stream.Cancel();
+                    }
+
+                    OperationCanceledException oce = e as OperationCanceledException;
```
You can delete this line and just change the above else if to:
```C#
else if (e is OperationCanceledException oce)
```

rmkerr replied:
I was't aware of that functionality. Thanks.
